common/peerprovider/ringpopprovider/provider.go

```diff
@@ -0,0 +1,208 @@
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package ringpopprovider
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"go.uber.org/yarpc/transport/tchannel"
+
+	"github.com/uber/ringpop-go"
+	"github.com/uber/ringpop-go/events"
+	"github.com/uber/ringpop-go/swim"
+	tcg "github.com/uber/tchannel-go"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/membership"
+)
+
+type (
+	// Provider use ringpop to announce membership changes
+	Provider struct {
+		status     int32
+		service    string
+		ringpop    *ringpop.Ringpop
+		bootParams *swim.BootstrapOptions
+		logger     log.Logger
+
+		smu         sync.RWMutex
+		subscribers map[string]chan<- *membership.ChangedEvent
+	}
+)
+
+const (
+	// roleKey label is set by every single service as soon as it bootstraps its
+	// ringpop instance. The data for this key is the service name
+	roleKey = "serviceName"
+)
+
+var _ membership.PeerProvider = (*Provider)(nil)
+
+func New(
+	service string,
+	config *Config,
+	channel tchannel.Channel,
+	logger log.Logger,
+) (*Provider, error) {
+	if err := config.validate(); err != nil {
+		return nil, err
+	}
+
+	rp, err := ringpop.New(config.Name, ringpop.Channel(channel.(*tcg.Channel)))
+	if err != nil {
+		return nil, fmt.Errorf("ringpop creation failed: %v", err)
+	}
+
+	discoveryProvider, err := newDiscoveryProvider(config, logger)
+	if err != nil {
+		return nil, fmt.Errorf("ringpop discovery provider: %v", err)
+	}
+
+	bootstrapOpts := &swim.BootstrapOptions{
+		MaxJoinDuration:  config.MaxJoinDuration,
+		DiscoverProvider: discoveryProvider,
+	}
+
+	return NewRingpopProvider(service, rp, bootstrapOpts, logger), nil
+}
+
+// NewRingpopProvider sets up ringpop based peer provider
+func NewRingpopProvider(service string,
+	rp *ringpop.Ringpop,
+	bootstrapOpts *swim.BootstrapOptions,
+	logger log.Logger,
+) *Provider {
+	return &Provider{
+		service:     service,
+		status:      common.DaemonStatusInitialized,
+		ringpop:     rp,
+		bootParams:  bootstrapOpts,
+		logger:      logger,
+		subscribers: map[string]chan<- *membership.ChangedEvent{},
+	}
+}
+
+// Start starts ringpop
+func (r *Provider) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&r.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	_, err := r.ringpop.Bootstrap(r.bootParams)
```

Comment:
what's the other value?  the bootstrapped ring or something?

Reply:
this is a list of nodes joined.